action.yml

```diff
@@ -26,6 +26,196 @@ inputs:
+        DOCKER_IMAGE_URL="${DOCKER_REGISTRY}/${DOCKER_REPO}:${DOCKER_TAG}"
+
+        # Modeled after: https://github.com/actions/runner/blob/1ceb1a67f270537fe80da23445fa78bfb97503ea/src/Runner.Worker/Container/ContainerInfo.cs#L37
+        # example: ghcrioactionsjekyllbuildpagesv107_7ca21f
+        DOCKER_CONTAINER_BASENAME=$(echo "${DOCKER_IMAGE_URL}" | sed 's/[^a-zA-Z0-9]//g')
+        NEW_GUID=$(uuidgen)
+        DOCKER_CONTAINER_NAME="${DOCKER_CONTAINER_BASENAME}_${NEW_GUID:0:6}"
+
+        # actions/runner source: https://github.com/actions/runner/blob/0484afeec71b612022e35ba80e5fe98a99cd0be8/src/Runner.Worker/Container/DockerCommandManager.cs#L49
+        # actions/runner example: 6c0442
+        # However, here, we're generating a SHA from the workflow run information
+        # instead of trying to figure out the equivalent directory path
+        LABEL_BASE="${GITHUB_RUN_ID}/${GITHUB_RUN_NUMBER}/${GITHUB_JOB}"
+        LABEL_SHA=$(echo "${LABEL_BASE}" | shasum -a 256 | awk '{print $1}')
+        DOCKER_CONTAINER_LABEL="${LABEL_SHA:0:6}"
+
+        # Craft an expected image name in case we need to do the local build
+        DOCKER_IMAGE_NAME=$DOCKER_IMAGE_URL
+        if [[ "$MUST_BUILD" == "true" ]]; then
+          DOCKER_IMAGE_NAME="${DOCKER_CONTAINER_LABEL}:$(uuidgen)"
+        fi
+
+        # actions/runner source: https://github.com/actions/runner/blob/0484afeec71b612022e35ba80e5fe98a99cd0be8/src/Runner.Worker/Container/DockerCommandManager.cs#L130-L140
+        # Gather list of environment variable names starting with GITHUB_
+        GITHUB_ENV_VARS=$(env | while IFS= read -r line; do key=${line%%=*}; if [[ "$key" =~ ^GITHUB_.+$ ]]; then printf -- '-e %s ' "$key"; fi; done)
+        # Gather list of environment variable names starting with ACTIONS_
+        ACTIONS_ENV_VARS=$(env | while IFS= read -r line; do key=${line%%=*}; if [[ "$key" =~ ^ACTIONS_.+$ ]]; then printf -- '-e %s ' "$key"; fi; done)
+        # Gather list of environment variable names with values starting with INPUT_
+        INPUT_ENV_VARS=$(env | while IFS= read -r line; do key=${line%%=*}; val=${line#*=}; if [[ "$key" =~ ^INPUT_.+$ ]]; then printf -- '-e %s=%s ' "$key" "$val"; fi; done)
+        PASSTHROUGH_ENV_VARS="${GITHUB_ENV_VARS}${ACTIONS_ENV_VARS}${INPUT_ENV_VARS}"
+
+        # Set step outputs
+        echo "building=${MUST_BUILD}" >> $GITHUB_OUTPUT
+        echo "image_url=${DOCKER_IMAGE_URL}" >> $GITHUB_OUTPUT
+        echo "image_name=${DOCKER_IMAGE_NAME}" >> $GITHUB_OUTPUT
+        echo "container_name=${DOCKER_CONTAINER_NAME}" >> $GITHUB_OUTPUT
+        echo "container_label=${DOCKER_CONTAINER_LABEL}" >> $GITHUB_OUTPUT
+        echo "env_expansion=${PASSTHROUGH_ENV_VARS}" >> $GITHUB_OUTPUT
+
+    - if: ${{ steps.docker.outputs.building != 'true' }}
+      name: Pull prebuilt Docker image
+      shell: sh
+      env:
+        DOCKER_IMAGE_URL: ${{ steps.docker.outputs.image_url }}
+      run: |
+        docker pull $DOCKER_IMAGE_URL
```

Comment:
If this step failed, that's when we would want to use the fallback URL

Reply:
I was wondering about that as well. 🤔 

My expectation was that if the `docker login` step succeeded, we would probably be good to go for pulling the image. But yeah, I'm OK with it falling back still after the `docker pull` if that's where we're going to see the friction.

My minor concern with that was accidentally allowing runs to pass when they might be attempting to pull a non-existent version, as that may silently lead us into an anti-pattern where we're building locally way more frequently than expected. 🤷🏻

Reply:
Confirmed with the GitHub Packages team that it is indeed possible to have a `docker login` succeed but a subsequent `docker pull` fail unexpectedly during some types of incidents. 😓

Reply:
Updated to cover this. 👍🏻